### ISSUE TITLE
SSH:// url clone migration feature

### ISSUE THREAD

**Issue by techzilla** [gogits/gogs#1565]
Wanted to request the ability to migrate/mirror git repositories that are only available via SSH.

System side would be ok, would mean a key would be needed in .ssh/id_rsa.  But then the other party would be responsible, and it would be a huge management reduction using gogs.

**Comment by Unknwon** [comment]
yeah... hmm.. thanks your feedback!

**Comment by NiklasRosenstein** [comment]
+1

**Comment by Dortamur** [comment]
As mentioned on the Discuss site (https://discuss.gogs.io/t/mirroring-from-an-ssh-repo-source/240/3) this can already work, mostly. You can manually do it by:

 * Set up an ssh key for the gogs user, to access the remote repo
 * Create a migration/mirror repo, using a local empty repo (eg; `mkdir /tmp/dummy; cd /tmp/dummy; git init` then use `/tmp/dummy`)
 * Go into the repo settings, and change the "Mirror Address" to your `gitolite:repo-name` real source via ssh

The main issue seems to just be URL validation blocking it (and the effort of manually setting up the local ssh config entry).

While the above trick works, I have a large number of repositories to migrate, which I'd really like to automate. I've been running some tests with the Gogs API, and if I do a new Migration via `POST /repos/migrate`, it has the same problem as the UI. I get the response:

```
{ status: 422,
  data: '{"message":"Invalid local path, it does not exist or not a directory.",
      "url":"https://github.com/gogits/go-gogs-client/wiki"}' }
```

(Running Gogs 0.9.13 on Debian 8.4)

**Comment by thincal** [comment]
+1

**Comment by thincal** [comment]
@Unknwon what's the status about this feature request ? thanks

**Comment by Unknwon** [comment]
@thincal no status...

**Comment by chill1n** [comment]
To support ssh:// what about the following changes:
1. Add check for "ssh://" prefix to modules/auth/repo_form.go: ParseRemoveAddr()
2. Add server ssh key to known_hosts file in routers/repo/repo.go: MigratePost(), e.g.
    ssh-keyscan -p <port> <remoteAddr> >> ~/.ssh/known_hosts

Handling a non-default ssh port would require separate parsing and a separate field to denote any non-default port number.

**Comment by bkcsoft** [comment]
If anyone wants to implement this: please **don't** use a system-wide ssh-key... That is a security nightmare...
(Setup and) Use a per-user key and pass that to `git clone`...

**Comment by sapk** [comment]
Maybe a good security practice could be that gogs can generate a uniq key at the mirroring configuration and display the public key to add to the distant repo ? That way the private key is manage by gogs and only ok for this distant repo ?

**Comment by bkcsoft** [comment]
@sapk Obviously, if/when this is implemented, Gogs will generate&track these keys :wink: 

My take on it is that each User has a set of keys (generated or manually added) that Gogs can use for different things around the system (think Deploy-keys, but in reverse). So when you add a ssh-migration you pick one of the keys, that you have, and Gogs will use that one for cloning it.

Could also be used for allowing CI/CD-runners to pull private repos.

**Comment by dantewang** [comment]
+1

**Comment by ziozzang** [comment]
+1

**Comment by Unknwon** [comment]
Possible solution: http://stackoverflow.com/questions/4565700/specify-private-ssh-key-to-use-when-executing-shell-command-with-or-without-ruby

**Comment by hsk81** [comment]
@Unknwon What's the status of this issue? Exactly the same problem here, where I want to migrate to Gogs -- great product! -- from gitolite.

**Comment by 275761919** [comment]
@Unknwon what's the status about this feature request ? thanks

**Comment by Unknwon** [comment]
@275761919 I'm concerned about security of this feature. If I understand correctly, you will need add private key to Gogs in order to make SSH clone work. That's something, doesn't feel right?

**Comment by 275761919** [comment]
@Unknwon An ssh key only for gogs can be generated.

**Comment by Unknwon** [comment]
@275761919 ah, good point.

**Comment by gotgot1995** [comment]
Same need here. I need to migrate one ton of legacy git repos to Gogs and I don't see myself manually copying all the repos and adding an origin to each and every one of them.

**Comment by biji** [comment]
> @275761919 I'm concerned about security of this feature. If I understand correctly, you will need add private key to Gogs in order to make SSH clone work. That's something, doesn't feel right?

we can run `ssh-keygen` as `git` user , then add public key in `/home/git/.ssh/id_rsa.pub` to github or bitbucket

**Comment by biji** [comment]
Minimum changes should be:
- Allow to enter URL like git@bitbucket.org:username/project.git in **New Migration**
Currently we got: *You are not allowed to import local repositories.*

or 

1. Create blank repository from **New Repository**
1. Allow to change repository as mirror in **Settings**

Thanks

**Comment by cicna** [comment]
git clone ssh://git@62.69.6.13:10022/abc/ying.git
